clinica/pipelines/engine.py

```diff
@@ -100,26 +118,71 @@ def __init__(self, bids_directory=None, caps_directory=None, tsv_file=None, name
                     (Fore.RED, self._name, Fore.RESET))
+
             check_caps_folder(self._caps_directory)
-            self._sessions, self._subjects = get_subject_session_list(
-                input_dir=self._caps_directory,
-                ss_file=self._tsv_file,
-                is_bids_dir=False
-            )
+            input_dir = self._caps_directory
+            is_bids_dir = True
         else:
             check_bids_folder(self._bids_directory)
-            self._sessions, self._subjects = get_subject_session_list(
-                input_dir=self._bids_directory,
-                ss_file=self._tsv_file,
-                is_bids_dir=True
-            )
+            input_dir = self._bids_directory
+            is_bids_dir = True
+        self.get_subject_session_list(input_dir,
+                                      self._tsv_file,
+                                      is_bids_dir,
+                                      self.base_dir)
 
         self.init_nodes()
 
-    def init_nodes(self):
-        """Init the basic workflow and I/O nodes necessary before build.
+    @staticmethod
+    def get_processed_images(caps_directory, subjects, sessions):
+        """Extract processed image IDs in `caps_directory` based on `subjects`_`sessions`.
 
+        Todo:
+            [ ] Implement this static method in all pipelines
+            [ ] Make it abstract to force overload in future pipelines
+        """
+        from clinica.utils.exceptions import ClinicaException
+        import datetime
+        from colorama import Fore
+        from clinica.utils.stream import cprint
+        now = datetime.datetime.now().strftime('%H:%M:%S')
+        cprint('\n%s[%s] Pipeline finished with errors.%s\n' % (Fore.RED, now, Fore.RESET))
+        cprint('%sCAPS outputs were not found for some image(s):%s' % (Fore.RED, Fore.RESET))
+        raise ClinicaException('Implementation on which image(s) failed will appear soon.')
+
+    def get_subject_session_list(self, input_dir, tsv_file, is_bids_dir, base_dir):
+        """Parse a BIDS or CAPS directory to get the subjects and sessions.
+
+        This function lists all the subjects and sessions based on the content of
+        the BIDS or CAPS directory or (if specified) on the provided
+        subject-sessions TSV file.
+
+        Notes:
+            This is a generic method based on folder names. If your <BIDS> dataset contains e.g.:
+            - sub-CLNC01/ses-M00/anat/sub-CLNC01_ses-M00_T1w.nii
+            - sub-CLNC02/ses-M00/dwi/sub-CLNC02_ses-M00_dwi.{bval|bvec|json|nii}
+            - sub-CLNC02/ses-M00/anat/sub-CLNC02_ses-M00_T1w.nii
+            get_subject_session_list(<BIDS>, None, True) will return
+            ['ses-M00', 'ses-M00'], ['sub-CLNC01', 'sub-CLNC02'].
+
+            However, if your pipeline needs both T1w and DWI files, you will need to overload this method.
```

Comment:
I agree that my first proposition would be very time consuming to modifiy all the pipelines. It is now removed in commit b5c80f15.

I think that we will eventually provide an iotool to generate TSV based on modalities on BIDS folder (e.g. `--modalities "T1w PET"`) and/or processed images in CAPS folder (e.g. `--pipelines t1-freesurfer`).

This would be useful for us when working with datasets having missing files e.g. ADNI: we could simply have TSV files describing having both T1w and PET, only T1w, both T1w and DWI, etc.

And for any pipeline, we would simply write the command the user will have to type e.g. `clinica iotools generate-tsv <BIDS> --modalities "T1w"` for `t1-freesurfer`. 

Anyway, we are not running out of ideas if we can add this feature :-)